Branch-SDK/src/main/java/io/branch/referral/ServerRequestRegisterOpen.java

```diff
@@ -46,7 +46,9 @@ public void onPreExecute() {
         super.onPreExecute();
         // Instant Deep Link if possible
         if (Branch.getInstance().isInstantDeepLinkPossible()) {
-            callback_.onInitFinished(Branch.getInstance().getLatestReferringParams(), null);
+            if (hasCallBack()) {
```

Comment:
Question:  Calling the method `hasCallback()` and then using the `callback_` member is non-optimal.   Can we instead have a stub (non-null) callback that checks to see if the caller passed one in, and call it if necessary?   It's a small thing...   but it turns out that every request class has the previous way and not all of them were changed.

Reply:
> but it turns out that every request class has the previous way and not all of them were changed.

Hmm.. every request I that see has my implementation, where we first check if callback is null (aka we call hasCallBack) and then we either invoke the callback or not.
See the implementation in:
 - ServerRequestRedeemRewards
 - ServerRequestRegisterInstall
 - ServerRequestGetRewards
 - ServerRequestIdentifyUserRequest
 - ServerRequestGetLATD
 - ServerRequestGetCPID (this one actually has the potential to throw an NPE, created SDK-756)

But if we were gonna go ahead with implementation you are describing, I fail to see how it is more optimal.. Do I understand you right (see below)?
```
Request {
    @NonNull StubCallback stubCallback = new StubCallback();
    Request(@Nullable BranchReferralInitListener c) {
        this.stubCallback = new StubCallback(c);
    }
    ...
    @Override public void onRequestSucceeded() {
        stubCallback.onInitFinished()
    }
    ...
    public static class StubCallback extends BranchReferralInitListener {
        private BranchReferralInitListener internalCallback;
        public StubCallback(@Nullable BranchReferralInitListener callback) {
            this.internalCallback = callback;
        }
       @Override public void onInitFinished() {
           if (internalCallback != null) {
               internalCallback.onInitFinished();
           }
       }
    }
}
```

Reply:
I am objecting only to the introduction of `hasCallback()` -- checking the local variable in that method, then using the local variable directly.   There is no advantage there (and is different than everywhere else) where you just say
```
if (callback_ != null) {
  callback_.doSomething();
}
```

However, if you *are* going to encapsulate the existence of a callback, then go all the way by making a stub and having the logic in there.   The advantage of doing that is that all of the other things that you do internally would be done in the stub as well.

Reply:
Consider these two patterns:
```
if (callback_ != null) {
  callback_.doSomething();
}
```

and

```
if (hasCallback()) {
  callback_.doSomething();
}
```

I maintain the second one is not a good pattern.

Reply:
Gotcha, that makes sense, I will revert to `callback_ != null`